File: packages/igloo-ts/src/widgets/row.ts
import { Row as WitRow } from 'iced:app/row@0.1.0';
import type { Length } from 'iced:app/length@0.1.0';
import type { Padding } from 'iced:app/padding@0.1.0';
import type { Vertical } from 'iced:app/alignment@0.1.0';
import type { Pixels } from 'iced:app/shared@0.1.0';
import { Element, toElement, type IntoElement, type ElementLike } from '../element.js';

/**
 * Builder for creating Row layout widgets.
 * A Row arranges its children horizontally.
 *
 * @example
 * ```typescript
 * const row = Row.new()
 *   .spacing(10)
 *   .push(Text.new('Left'))
 *   .push(Text.new('Right'));
 * ```
 *
 * The child message type is inferred and accumulated as elements are pushed;
 * pass `Row.new<Msg>()` to pin it so each `push` is checked against `Msg`.
 *
 * @typeParam Msg - The accumulated message type of the row's children.
 * @typeParam Bound - Upper bound each child must satisfy; `unknown` when unpinned.
 */
export class Row<Msg = never, Bound = unknown> implements IntoElement<Msg> {
  private raw: WitRow;

  private constructor() {
    this.raw = new WitRow();
  }

  /** Create a new empty Row builder */
  static new(): Row<never, unknown>;
  /** Create a new empty Row builder pinned to the message type `B` */
  static new<B>(): Row<never, B>;
  static new(): Row<never, unknown> {
    return new Row();
  }

  /** Create a Row with the given elements */
  static with<const M>(elements: ElementLike<M>[]): Row<M, unknown> {
    return new Row<M, unknown>().extend(elements);
  }

  /** Add an element to the row */
  push<const M extends Bound>(element: ElementLike<M>): Row<Msg | M, Bound> {
    this.raw.push(toElement(element).inner);
    return this as unknown as Row<Msg | M, Bound>;
  }

  /** Add an element conditionally */
  pushIf<const M extends Bound>(
    condition: boolean,
    element: () => ElementLike<M>,
  ): Row<Msg | M, Bound> {
    if (condition) {
      this.raw.push(toElement(element()).inner);
    }
    return this as unknown as Row<Msg | M, Bound>;
  }

  /** Add multiple elements */
  extend<const M extends Bound>(elements: ElementLike<M>[]): Row<Msg | M, Bound> {
    for (const element of elements) {
      this.raw.push(toElement(element).inner);
    }
    return this as unknown as Row<Msg | M, Bound>;
  }

  /** Set the spacing between elements in pixels */
  spacing(spacing: Pixels): this {
    this.raw.spacing(spacing);
    return this;
  }

  /** Set the padding around the row */
  padding(padding: Padding): this {
    this.raw.padding(padding);
    return this;
  }

  /** Set the row width */
  width(width: Length): this {
    this.raw.width(width);
    return this;
  }

  /** Set the row height */
  height(height: Length): this {
    this.raw.height(height);
    return this;
  }

  /** Set vertical alignment of children */
  alignY(align: Vertical): this {
    this.raw.alignY(align);
    return this;
  }

  /** Enable or disable clipping of content */
  clip(clip: boolean = true): this {
    this.raw.clip(clip);
    return this;
  }

  /** Enable or disable wrapping of elements to the next line */
  wrap(wrap: boolean = true): this {
    this.raw.wrap(wrap);
    return this;
  }

  /** Convert to Element (implements IntoElement) */
  intoElement(): Element<Msg> {
    return new Element(WitRow.intoElement(this.raw));
  }
}

File: packages/igloo-py/src/igloo_py/widgets/row.py
"""Row layout widget builder."""

from collections.abc import Callable
from typing import TYPE_CHECKING

from wit_world.imports.row import Row as WitRow

from ..element import Element, ElementLike, IntoElement, to_element

if TYPE_CHECKING:
    from ..types.alignment import WitVertical
    from ..types.length import WitLength
    from ..types.padding import WitPadding


class Row(IntoElement):
    """
    Builder for creating Row layout widgets.
    A Row arranges its children horizontally.

    Example:
        row = Row.new()
            .spacing(10)
            .push(Text.new("Left"))
            .push(Text.new("Right"))
    """

    def __init__(self) -> None:
        self._raw = WitRow()

    @classmethod
    def new(cls) -> "Row":
        """Create a new empty Row builder."""
        return cls()

    @classmethod
    def with_elements(cls, elements: list[ElementLike]) -> "Row":
        """Create a Row with the given elements."""
        row = cls()
        row.extend(elements)
        return row

    def push(self, element: ElementLike) -> "Row":
        """Add an element to the row."""
        self._raw.push(to_element(element).inner)
        return self

    def push_if(self, condition: bool, element: Callable[[], ElementLike]) -> "Row":
        """Add an element conditionally."""
        if condition:
            self._raw.push(to_element(element()).inner)
        return self

    def extend(self, elements: list[ElementLike]) -> "Row":
        """Add multiple elements."""
        for element in elements:
            self._raw.push(to_element(element).inner)
        return self

    def spacing(self, spacing: float) -> "Row":
        """Set the spacing between elements in pixels."""
        self._raw.spacing(spacing)
        return self

    def padding(self, padding: "WitPadding") -> "Row":
        """Set the padding around the row."""
        self._raw.padding(padding)
        return self

    def width(self, width: "WitLength") -> "Row":
        """Set the row width."""
        self._raw.width(width)
        return self

    def height(self, height: "WitLength") -> "Row":
        """Set the row height."""
        self._raw.height(height)
        return self

    def align_y(self, align: "WitVertical") -> "Row":
        """Set vertical alignment of children."""
        self._raw.align_y(align)
        return self

    def clip(self, clip: bool = True) -> "Row":
        """Enable or disable clipping of content."""
        self._raw.clip(clip)
        return self

    def wrap(self, wrap: bool = True) -> "Row":
        """Enable or disable wrapping of elements to the next line."""
        self._raw.wrap(wrap)
        return self

    def into_element(self) -> Element:
        """Convert to Element (implements IntoElement)."""
        return Element(WitRow.into_element(self._raw))
